drivers/usb/dwc3/core.c

```diff
@@ -311,9 +311,46 @@ int dwc3_core_soft_reset(struct dwc3 *dwc)
 	 * XHCI driver will reset the host block. If dwc3 was configured for
 	 * host-only mode or current role is host, then we can return early.
 	 */
-	if (dwc->dr_mode == USB_DR_MODE_HOST || dwc->current_dr_role == DWC3_GCTL_PRTCAP_HOST)
+	if (dwc->current_dr_role == DWC3_GCTL_PRTCAP_HOST)
 		return 0;
 
+	/*
+	 * If the dr_mode is host and the dwc->current_dr_role is not the
+	 * corresponding DWC3_GCTL_PRTCAP_HOST, then the dwc3_core_init_mode
+	 * isn't executed yet. Ensure the phy is ready before the controller
+	 * updates the GCTL.PRTCAPDIR or other settings by soft-resetting
+	 * the phy.
+	 *
+	 * Note: GUSB3PIPECTL[n] and GUSB2PHYCFG[n] are port settings where n
+	 * is port index. If this is a multiport host, then we need to reset
+	 * all active ports.
+	 */
+	if (dwc->dr_mode == USB_DR_MODE_HOST) {
+		u32 usb3_port;
+		u32 usb2_port;
+
+		usb3_port = dwc3_readl(dwc->regs, DWC3_GUSB3PIPECTL(0));
+		usb3_port |= DWC3_GUSB3PIPECTL_PHYSOFTRST;
+		dwc3_writel(dwc->regs, DWC3_GUSB3PIPECTL(0), usb3_port);
+
+		usb2_port = dwc3_readl(dwc->regs, DWC3_GUSB2PHYCFG(0));
+		usb2_port |= DWC3_GUSB2PHYCFG_PHYSOFTRST;
+		dwc3_writel(dwc->regs, DWC3_GUSB2PHYCFG(0), usb2_port);
+
+		/* Small delay for phy reset assertion */
+		usleep_range(1000, 2000);
+
+		usb3_port &= ~DWC3_GUSB3PIPECTL_PHYSOFTRST;
+		dwc3_writel(dwc->regs, DWC3_GUSB3PIPECTL(0), usb3_port);
+
+		usb2_port &= ~DWC3_GUSB2PHYCFG_PHYSOFTRST;
+		dwc3_writel(dwc->regs, DWC3_GUSB2PHYCFG(0), usb2_port);
+
+		/* Wait for clock synchronization */
+		msleep(50);
+		return 0;
+	}
+
 	if (dwc->has_dsm_for_softreset) {
```

Comment:
You resolved this merge differently than @mike-petersen-ni did in #138; he put the `if (dwc->has_dsm_for_softreset)` conditional before the new `if (dwc->dr_mode == USB_DR_MODE_HOST)` check.

I think the way you have it might be the correct one: we had `if (dwc->has_dsm_for_softreset)` after:
```c
if (dwc->dr_mode == USB_DR_MODE_HOST || dwc->current_dr_role == DWC3_GCTL_PRTCAP_HOST)
    return 0;
```
and then upstream split that into
```c
if (dwc->current_dr_role == DWC3_GCTL_PRTCAP_HOST)
    return 0;
if (dwc->dr_mode == USB_DR_MODE_HOST) {
   /*do a bunch of stuff*/ 
   return 0;
}
```
and so I think `if (dwc->has_dsm_for_softreset)` should still go after that whole block?

I'm in agreement with @gratian though on the subject of the patch we're carrying; it's an out-of-tree thing we got out of a Yocto tree and doesn't seem to have made it upstream at all, and I also wonder if it's still necessary.

Reply:
That was my rationale for positioning, yea. A previous conditional with an OR was simply split in two. Do you agree (given your approval) that considering dropping our patch is out of scope of this merge? Ideally it'd be dropped explicitly, or as part of the rebase?

Reply:
Yeah, dropping it is out-of-scope for the merge; we'd want to drop it explicitly (after doing proper testing to see if we still need it or not)